### PR TITLE
Fc 816 display training provider locations on employer favourites

### PR DESCRIPTION
**Pull request by ben1stone-leftdfe**

### CODE REVIEW

File: src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs [hunk 1]
```diff
@@ -118,7 +118,10 @@ private ApprenticeshipFavourites MapToWriteModel(List<Favourite> favourites)
             }
 
             var items = favourites.Select(x => new ApprenticeshipFavourite
-            { ApprenticeshipId = x.ApprenticeshipId, Ukprns = x.Ukprns });
+            {
+                ApprenticeshipId = x.ApprenticeshipId, Providers = x.Providers
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs]: Each property on a new line to be consistent.

File: src/web/EmployerFavouritesWeb/EmployerFavouritesWeb.csproj [hunk 2]
```diff
@@ -21,6 +21,7 @@
     <PackageReference Include="Microsoft.Azure.Cosmos.Table" Version="1.0.1" />
     <PackageReference Include="SFA.DAS.Account.Api.Client" Version="1.6.1" />
     <PackageReference Include="Polly" Version="7.1.0" />
+    <PackageReference Include="SFA.DAS.SAS.Shared.Components" Version="1.1.0.14-alpha-basket" />
```
Comment (leewadhams) [inline src/web/EmployerFavouritesWeb/EmployerFavouritesWeb.csproj]: Don't think this is required is it?

File: src/api/EmployerFavouritesApi/Domain/ReadModel/Provider.cs [hunk 3]
```diff
@@ -1,8 +1,13 @@
...
+using System.Collections.Generic;
+
+namespace DfE.EmployerFavourites.Api.Domain.ReadModel
 {
     public class Provider
     {
         public int Ukprn { get; set; }
         public string Name { get; set; }
+        public List<int> LocationIds { get; set; }
+        public List<Location> Locations { get; set; }
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Domain/ReadModel/Provider.cs]: Do we need both of these properties?

File: src/api/EmployerFavouritesApi/Domain/WriteModel/ApprenticeshipFavourite.cs [hunk 4]
```diff
@@ -1,15 +1,18 @@
...
         public ApprenticeshipFavourite()
         {
-            Ukprns = new List<int>();
+            Providers = new List<Provider>();
         }
 
         public string ApprenticeshipId { get; set; }
-        public IList<int> Ukprns { get; set; }
+        public IList<Provider> Providers { get; set; }
+
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Domain/WriteModel/ApprenticeshipFavourite.cs]: Blank lines

File: src/api/EmployerFavouritesApi/Infrastructure/Interfaces/IFatRepository.cs [hunk 5]
```diff
@@ -1,3 +1,5 @@
+using DfE.EmployerFavourites.Api.Models;
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Infrastructure/Interfaces/IFatRepository.cs]: Usings not required

File: src/api/EmployerFavouritesApi/Domain/WriteModel/ApprenticeshipFavourites.cs [hunk 6]
```diff
@@ -13,7 +13,9 @@ public void Remove(string apprenticeshipId)
 
         public void Remove(string apprenticeshipId, int ukprn)
         {
-            this.Where(w => w.ApprenticeshipId == apprenticeshipId).ToList().ForEach(item => item.Ukprns.Remove(ukprn));
+            var provider = this.FirstOrDefault(w => w.ApprenticeshipId == apprenticeshipId)?.Providers;
+
+            provider.Remove(provider.SingleOrDefault(w => w.Ukprn == ukprn));
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Domain/WriteModel/ApprenticeshipFavourites.cs]: Probably need a null check on provider as the last line will fail
```
if (provider != null)
    provider.Remove(provider.SingleOrDefault(w => w.Ukprn == ukprn));
```

File: src/api/EmployerFavouritesApi/Infrastructure/FatApiRepository.cs [hunk 7]
```diff
@@ -1,8 +1,11 @@
 ﻿using System;
+using System.Collections.Generic;
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Infrastructure/FatApiRepository.cs]: unused usings

File: src/api/EmployerFavouritesApi/Infrastructure/FatFavouritesTableStorageRepository.cs [hunk 8]
```diff
@@ -93,11 +95,12 @@ private async Task<Domain.ReadModel.ApprenticeshipFavourite> BuildReadModelItem(
...
-                Providers = await Task.WhenAll(src.Ukprns?.Select(async x => new Domain.ReadModel.Provider
+                Providers = await Task.WhenAll(src.Providers?.Select(async x => new Domain.ReadModel.Provider
                 {
-                    Ukprn = x,
-                    Name = await _fatRepository.GetProviderNameAsync(x)
-
+                    Ukprn = x.Ukprn,
+                    Name = await _fatRepository.GetProviderNameAsync(x.Ukprn),
+                    LocationIds = x.LocationIds,
+                    Locations = new List<Location>()
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Infrastructure/FatFavouritesTableStorageRepository.cs]: Same question as before. Do we need both properties?

File: src/api/EmployerFavouritesApi/Domain/ReadModel/ApprenticeshipFavourites.cs [hunk 9]
```diff
@@ -12,7 +13,7 @@ public WriteModel.ApprenticeshipFavourites MapToWriteModel()
             model.AddRange(this.Select(x => new WriteModel.ApprenticeshipFavourite
             {
                 ApprenticeshipId = x.ApprenticeshipId,
-                Ukprns = x?.Providers.Select(y => y.Ukprn).ToList() ?? new List<int>(0)
+                Providers = x?.Providers.Select(y => new Models.Provider { Ukprn = y.Ukprn, LocationIds = y.LocationIds }).ToList() ?? new List<Models.Provider>()
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Domain/ReadModel/ApprenticeshipFavourites.cs]: The Providers type shouldn't be using the Models.Provider. The models providers are used in the UI whereas this class is a domain model (centre of the onion). Granted it's not that easy to see from the folders. It's maybe time we split the project into the separate layers to make things clearer.

File: src/api/EmployerFavouritesApi/Models/Location.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
+﻿using DfE.EmployerFavourites.Api.Infrastructure;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace DfE.EmployerFavourites.Api.Models
+{
+    public class Location
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Models/Location.cs:9]: Following from the previous comment a class for representing Location would need to exist in the Domain.

File: src/api/EmployerFavouritesApi/Models/Provider.cs [hunk 11]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace DfE.EmployerFavourites.Api.Models
+{
+    public class Provider
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Models/Provider.cs:8]: Following from the previous comment a class for representing Provider would need to exist in the Domain.

File: src/api/UnitTests/Domain/ReadModel/ApprenticeshipFavouritesTests.cs [hunk 12]
```diff
@@ -24,51 +25,34 @@ public void MapToWriteModel_ReturnWriteModelWithPropertiesPopulated()
             Assert.NotNull(result);
             Assert.Equal(sut.Count, result.Count);
 
-            Assert.Collection(result,
-                item => { Assert.Equal("ABC123", item.ApprenticeshipId); Assert.Equal(new List<int> { 1, 2, 3 }, item.Ukprns); },
-                item => { Assert.Equal("XYZ123", item.ApprenticeshipId); Assert.Equal(0, item.Ukprns.Count); });
+            //Assert.Collection(result,
```
Comment (leewadhams) [inline src/api/UnitTests/Domain/ReadModel/ApprenticeshipFavouritesTests.cs]: Can the commented out code be deleted?

File: src/web/EmployerFavouritesWeb/Domain/ReadModel/ApprenticeshipFavourites.cs [hunk 13]
```diff
@@ -12,8 +12,8 @@ public WriteModel.ApprenticeshipFavourites MapToWriteModel()
             model.AddRange(this.Select(x => new Domain.WriteModel.ApprenticeshipFavourite
             {
                 ApprenticeshipId = x.ApprenticeshipId,
-                Ukprns = x?.Providers.Select(y => y.Ukprn).ToList() ?? new List<int>(0)
-            }));
+                Providers = x?.Providers.Select(y => new WriteModel.Provider(y.Ukprn, y.Locations )).ToList()  // ?? new List<Provider>()
```
Comment (leewadhams) [inline src/web/EmployerFavouritesWeb/Domain/ReadModel/ApprenticeshipFavourites.cs]: We need the commented out code don't we for a default?

File: src/web/EmployerFavouritesWeb/Domain/ReadModel/Provider.cs [hunk 14]
```diff
@@ -11,5 +13,8 @@ public class Provider
         public Uri Website { get; set; }
         public double EmployerSatisfaction { get; set; }
         public double LearnerSatisfaction { get; set; }
+        public List<Location> Locations { get; set; }
+        public ProviderAddress Address { get; set; }
+        public IList<int> LocationIds { get; set; }
```
Comment (leewadhams) [inline src/web/EmployerFavouritesWeb/Domain/ReadModel/Provider.cs:18]: Again we have 2 location lists. Do we need them both?

File: src/web/EmployerFavouritesWeb/Domain/WriteModel/ApprenticeshipFavourite.cs [hunk 15]
```diff
@@ -1,31 +1,32 @@
...
-                Ukprns = ukprns;
+            if (providers != null)
+                Providers = providers.Select(p => new Provider (p.Key,p.Value )).ToList();
         }
 
-        public ApprenticeshipFavourite(string apprenticeshipId, int ukprn) : this(apprenticeshipId)
-        {
-            Ukprns.Add(ukprn);
-        }
+        //public ApprenticeshipFavourite(string apprenticeshipId, Provider provider) : this(apprenticeshipId)
```
Comment (leewadhams) [inline src/web/EmployerFavouritesWeb/Domain/WriteModel/ApprenticeshipFavourite.cs]: Can this be removed?

File: src/web/EmployerFavouritesWeb/Models/JsonLocation.cs [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+namespace DfE.EmployerFavourites.Web.Models
+{
+    public class JsonLocation
+    {
+        //public int LocationId { get; set; }
+        //public string LocationName { get; set; }
+        //public JsonAddress Address { get; set; }
+        //public string Phone { get; set; }
+        //public string Email { get; set; }
+    }
```
Comment (leewadhams) [inline src/web/EmployerFavouritesWeb/Models/JsonLocation.cs]: Is this class needed?

File: src/web/UnitTests/Domain/ReadModel/ApprenticeshipFavouriteTests.cs [hunk 17]
```diff
@@ -39,5 +44,19 @@ public void IsFramework_ReturnFalse_WhenStandard(string id)
...
+        //    List<ProviderAddress> Address = new List<ProviderAddress> { new ProviderAddress { ContactType = "LEGAL" },
+        //                                                                new ProviderAddress { ContactType = "PRIMARY" } };
+        //    var generator = new ();
+        //}
+
+        [Fact]
+        public void GenerateAddress_WhenAllInformationIsAvailable()
+        {
+
+        }
```
Comment (leewadhams) [inline src/web/UnitTests/Domain/ReadModel/ApprenticeshipFavouriteTests.cs]: Is this yet to be implemented?

File: src/web/UnitTests/Domain/ReadModel/ApprenticeshipFavouritesTests.cs [hunk 18]
```diff
@@ -20,9 +20,9 @@ public void MapToWriteModel_ReturnWriteModelWithPropertiesPopulated()
             Assert.NotNull(result);
             Assert.Equal(sut.Count, result.Count);
 
-            Assert.Collection(result,
-                item => { Assert.Equal("ABC123", item.ApprenticeshipId); Assert.Equal(new List<int> { 1, 2, 3 }, item.Ukprns); },
-                item => { Assert.Equal("XYZ123", item.ApprenticeshipId); Assert.Equal(0, item.Ukprns.Count); });
+            //Assert.Collection(result,
+            //    item => { Assert.Equal("ABC123", item.ApprenticeshipId); Assert.Equal(new List<Provider> { new Provider { Ukprn = 1 }, new Provider { Ukprn = 2 }, new Provider { Ukprn = 3 } }, item.Providers); },
+            //    item => { Assert.Equal("XYZ123", item.ApprenticeshipId); Assert.Equal(0, item.Providers.Count); });
```
Comment (leewadhams) [inline src/web/UnitTests/Domain/ReadModel/ApprenticeshipFavouritesTests.cs]: Can this be deleted?

**Review by leewadhams — CHANGES_REQUESTED** [review]
Nice Ben 👍 

A few comments/tidy up things. Looking pretty good though mate.

File: src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs (hunk 1 above)
Comment (ben1stone-leftdfe) [reply src/api/EmployerFavouritesApi/Controllers/ApprenticeshipsController.cs]: Done

File: src/web/EmployerFavouritesWeb/EmployerFavouritesWeb.csproj (hunk 2 above)
Comment (ben1stone-leftdfe) [reply src/web/EmployerFavouritesWeb/EmployerFavouritesWeb.csproj]: Removed

File: src/api/EmployerFavouritesApi/Domain/WriteModel/ApprenticeshipFavourite.cs (hunk 4 above)
Comment (ben1stone-leftdfe) [reply src/api/EmployerFavouritesApi/Domain/WriteModel/ApprenticeshipFavourite.cs]: Removed

File: src/api/EmployerFavouritesApi/Infrastructure/Interfaces/IFatRepository.cs (hunk 5 above)
Comment (ben1stone-leftdfe) [reply src/api/EmployerFavouritesApi/Infrastructure/Interfaces/IFatRepository.cs]: Removed

File: src/api/EmployerFavouritesApi/Domain/WriteModel/ApprenticeshipFavourites.cs (hunk 6 above)
Comment (ben1stone-leftdfe) [reply src/api/EmployerFavouritesApi/Domain/WriteModel/ApprenticeshipFavourites.cs]: Done

File: src/api/EmployerFavouritesApi/Infrastructure/FatApiRepository.cs (hunk 7 above)
Comment (ben1stone-leftdfe) [reply src/api/EmployerFavouritesApi/Infrastructure/FatApiRepository.cs]: Removed

File: src/api/EmployerFavouritesApi/Domain/ReadModel/ApprenticeshipFavourites.cs (hunk 9 above)
Comment (ben1stone-leftdfe) [reply src/api/EmployerFavouritesApi/Domain/ReadModel/ApprenticeshipFavourites.cs]: Done

File: src/api/EmployerFavouritesApi/Models/Location.cs (hunk 10 above)
Comment (ben1stone-leftdfe) [reply src/api/EmployerFavouritesApi/Models/Location.cs:9]: Done

File: src/api/EmployerFavouritesApi/Models/Provider.cs (hunk 11 above)
Comment (ben1stone-leftdfe) [reply src/api/EmployerFavouritesApi/Models/Provider.cs:8]: Done

File: src/api/UnitTests/Domain/ReadModel/ApprenticeshipFavouritesTests.cs (hunk 12 above)
Comment (ben1stone-leftdfe) [reply src/api/UnitTests/Domain/ReadModel/ApprenticeshipFavouritesTests.cs]: This has been reworked.

File: src/web/EmployerFavouritesWeb/Domain/ReadModel/ApprenticeshipFavourites.cs (hunk 13 above)
Comment (ben1stone-leftdfe) [reply src/web/EmployerFavouritesWeb/Domain/ReadModel/ApprenticeshipFavourites.cs]: Sorted

File: src/web/EmployerFavouritesWeb/Domain/WriteModel/ApprenticeshipFavourite.cs (hunk 15 above)
Comment (ben1stone-leftdfe) [reply src/web/EmployerFavouritesWeb/Domain/WriteModel/ApprenticeshipFavourite.cs]: Yes, it was restructured to use a dictionary

File: src/web/EmployerFavouritesWeb/Models/JsonLocation.cs (hunk 16 above)
Comment (ben1stone-leftdfe) [reply src/web/EmployerFavouritesWeb/Models/JsonLocation.cs]: Removed

File: src/web/UnitTests/Domain/ReadModel/ApprenticeshipFavouriteTests.cs (hunk 17 above)
Comment (ben1stone-leftdfe) [reply src/web/UnitTests/Domain/ReadModel/ApprenticeshipFavouriteTests.cs]: This was something I was looking at implementing but now it maps each element of the address and joins them in the view model instead.

File: src/web/UnitTests/Domain/ReadModel/ApprenticeshipFavouritesTests.cs (hunk 18 above)
Comment (ben1stone-leftdfe) [reply src/web/UnitTests/Domain/ReadModel/ApprenticeshipFavouritesTests.cs]: Yes

**Comment by ben1stone-leftdfe** [comment]
I have completed all the changes

File: src/api/EmployerFavouritesApi/Infrastructure/FatFavouritesTableStorageRepository.cs (hunk 8 above)
Comment (ben1stone-leftdfe) [reply src/api/EmployerFavouritesApi/Infrastructure/FatFavouritesTableStorageRepository.cs]: LocationIds is used differently to the Locations list so we do need both. LocationIds is a list pulled from storage which is then used to populate the locations list.

File: src/api/EmployerFavouritesApi/Domain/ReadModel/Location.cs [hunk 19]
```diff
@@ -0,0 +1,20 @@
+﻿using DfE.EmployerFavourites.Api.Infrastructure;
```
Comment (leewadhams) [inline src/api/EmployerFavouritesApi/Domain/ReadModel/Location.cs]: Unused usings

**Review by leewadhams — APPROVED** [review]

**Pull request merged**